plugins/FirmwareUpdater/FirmwareUpdaterMachineAction.py

```diff
@@ -0,0 +1,68 @@
+# Copyright (c) 2018 Ultimaker B.V.
+# Cura is released under the terms of the LGPLv3 or higher.
+
+from cura.CuraApplication import CuraApplication
+from UM.Settings.DefinitionContainer import DefinitionContainer
+from cura.MachineAction import MachineAction
+from UM.i18n import i18nCatalog
+from UM.Settings.ContainerRegistry import ContainerRegistry
+from cura.PrinterOutput.FirmwareUpdater import FirmwareUpdateState
+
+from PyQt5.QtCore import pyqtSignal, pyqtProperty, QObject
+from typing import Optional
+
+MYPY = False
+if MYPY:
+    from cura.PrinterOutput.FirmwareUpdater import FirmwareUpdater
+    from cura.PrinterOutput.PrinterOutputDevice import PrinterOutputDevice
+
+catalog = i18nCatalog("cura")
+
+##  Upgrade the firmware of a machine by USB with this action.
+class FirmwareUpdaterMachineAction(MachineAction):
+    def __init__(self) -> None:
+        super().__init__("UpgradeFirmware", catalog.i18nc("@action", "Update Firmware"))
+        self._qml_url = "FirmwareUpdaterMachineAction.qml"
+        ContainerRegistry.getInstance().containerAdded.connect(self._onContainerAdded)
+
+        self._active_output_device = None #type: Optional[PrinterOutputDevice]
+        self._active_firmware_updater = None #type: Optional[FirmwareUpdater]
+
+        CuraApplication.getInstance().engineCreatedSignal.connect(self._onEngineCreated)
+
+    def _onEngineCreated(self) -> None:
+        CuraApplication.getInstance().getMachineManager().outputDevicesChanged.connect(self._onOutputDevicesChanged)
+
+    def _onContainerAdded(self, container) -> None:
```

Comment:
Missing type for container (I think that ContainerInterface should be sufficient here)